### ISSUE TITLE
[META] Ideas to improve CMI UX and make the feature usable by less technical people

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3475]
List of related issues:

- [ ] #408 - Export/Import an entire content type (or other entity type) with its fields all at once
- [ ] #661 - Config import should search recursively for config files to import
- [ ] #667 - Config: incorrect warning thrown when no files are found in import
- [ ] #1242 - Do not allow modules to be enabled at all if they have conflicting config file names with any existing config.
- [ ] #1654 - upload an CMI config.tar.gz automatically re: #1199
- [ ] #2236 - Config import cannot always name the needed module.
- [ ] #2263 - Ignore README.md when checking if active config directory is empty on install
- [ ] #2604 - List modules that need to be enabled on the config sync page 
- [ ] #3022 - All caches should be cleared after config sync
- [ ] #3112 - No error in admin UI when config management can't access staging directory
- [ ] #3119 - Consider using .zip instead of .tar for the config import/export
- [ ] #3839 - Show location of config files on status report
- [ ] #4188 - Allow config files to have dependencies
- [ ] #4189 - Add cron job to clean up configuration files for uninstalled contributed files
- [ ] #4212 - Keep a snapshot of configuration directory in database
- [ ] #4217 - Config import excludes last file in some imports - depending upon how tarball is created
- [ ] #4344 - Unable to config sync due to `reserved field names` false positive
- [x] #4447 - Provide advanced admin option to skip deleting config staging files after a config sync

------

These are a few suggestions/ideas made during last week's Virtual User Group meeting (which unfortunately could not be recorded due to technical issues):

- Allow multiple configuration sets to be selected for export. Perhaps convert the drop-down menus in `/admin/config/development/configuration/single/export` + offer a "download configuration" button which exports all .json config files.

- Have selection of config work like features (auto-selecting dependent configurations).

- Make configuration file names have more intuitive names and/or add human-readable names/descriptions in them.

- Consider providing means to version configuration, and then allow for configuration diff to be exported. Scenario: provide a way to "start recording" configuration changes from a specific point in time in local/staging ...make config changes ..."stop recording" configuration and export it. "Replay" the same configuration diff/scenario on prod.

These suggestions come from a non-technical point of view, so they might either be difficult to implement, or even not possible at all. They are ideas coming from people that find the whole concept and/or also the workflow of CMI hard to get their heads around.

**Comment by findlabnet** [comment]
IMHO, less technical people should do less technical things. 
Hope my opinion do not contrary to any CoC.

**Comment by docwilmot** [comment]
> IMHO, less technical people should do less technical things.

Then no one would learn.

**Comment by findlabnet** [comment]
> Then no one would learn.

Why?

**Comment by klonos** [comment]
@findlabnet I appreciate that this is your personal opinion. It is perfect timing that only a few hours ago I have posted a link to this article in our Gitter channel: https://www.drupal.org/forum/general/general-discussion/2010-11-05/why-is-drupal-always-screwed-up

...I stressed this bit in that article:

>I HATE DRUPAL WITH A PASSION. It just seems like a "Bunch of techs writing code and chuckling at all the "Low life non coders" who cant use it."

If that is not enough to make a point, then I am going to put my member of the Project Management Committee hat on, and simply say that I would like to point you to our official values: https://backdropcms.org/philosophy

...especially this (emphasis mine):

>It's most important that the product first be usable, and second, easy to build and extend. **Developers should consider the editorial experience first, the architect experience second, and put the needs of other developers last**. Intuitive user interfaces are more valuable than extensive flexibility. When options are presented in the user interface, they should be made as clear and concise as possible.

**Comment by klonos** [comment]
> Anyway, my hope for CMI was that (at a minimum) it would replace Features module;

You are not the only one @cellear. I was expecting the same till I actually got my head around how this works. This meta is a good start to try and flesh out pain points, and I think that making CMI work more like Features would be a good thing (additionally to its current implementation though - not change how it currently works). But I imagine that it will require quite some work to get it done.

**Comment by klonos** [comment]
TL;DR: the idea is to allow novices to set source/destination sites, and allow them to push config from on to the other. Read on...

I was watching @jenlampton's recent MiniCamp presentation on config management, trying to follow along, but wearing a non-coder/novice-user hat. I couldn't help thinking of possible ways for such users to avoid touching the command line and/or ftp. ...I then thought that we already have the UI bits in place for viewing changes in a nice color-coded diff, and then apply them; what is "missing" is the bit of how to get the configuration from one site (say local/stage) to another (live/prod).

...I then recalled https://www.drupal.org/project/deploy, which helps people "push" staged _content_ into another site, and though that there could be a similar solution that does the same, _but for config_.

...I then thought that perhaps this could be a security concern (man in the middle attacks when config is being pushed to a destination) ...but then also thought of #1992 and how we are working on implementing signatures for modules/projects. Could we use that same signing before we push the config, so to make it safer?

**Comment by jenlampton** [comment]
I like what you're cooking!

**Comment by stpaultim** [comment]
I just added a bunch of issues to this META. I'm not sure if all of these issues necessarily belong here, but lacking a better META issue, I started with this one.

**Comment by jenlampton** [comment]
I went one step further and created a Project for CMI issues: https://github.com/backdrop/backdrop-issues/projects/19#card-66338232 I've added everything in this META issue to the Project, so I don't think we need this issue anymore, but I've also added it to the project so that we can reference the comments from there if needed :)
